Feature/count votes

mattlee68 opened the pull request.

mattlee68 committed:
CC#4 Add admin dashboard page

mattlee68 committed:
CC#4 Add ability to vote using POST url

mattlee68 committed:
CC#4 Create use case pattern for Slack hook

mattlee68 committed:
CC#4 Attempt at using mocks and stubs for testing API

mattlee68 committed:
CC#4 Create test for sending slack notification

mattlee68 committed:
CC#8 Count votes and display on admin dashboard

mattlee68 committed:
CC#8 Count votes and display on admin dashboard

mattlee68 committed:
CC#8 Add tests for creating report use case

mattlee68 committed:
CC#8 Add tests for creating report use case

mattlee68 committed:
CC#8 Fix up multiple data queries and tidied up views

mattlee68 committed:
CC#8 Fix up multiple data queries and tidied up views

mattlee68 committed:
CC#8 Fix up tests to have correct fixtures

mattlee68 committed:
CC#8 Fix up tests to have correct fixtures

mattlee68 committed:
CC#8 Fix up tests to avoid race condition with Rspec and Fixtures

mattlee68 committed:
CC#8 Fix up tests to avoid race condition with Rspec and Fixtures

mattlee68 committed:
CC#8 Fix up create voting report test

mattlee68 committed:
CC#8 Fix up create voting report test

mattlee68 committed:
CC#4 Fix requested changes and add validation tests for Vote

mattlee68 committed:
CC#8 Fix authenticity token

mattlee68 committed:
CC#8 Fix authenticity token

marcusbaguley commented on app/use_cases/create_voting_report.rb:
```diff
@@ -0,0 +1,50 @@
+      date_before_next_batch = find_date_before_next_batch(batch)
+
+      all_votes_between_dates = votes.select do |vote|
+        vote.date >= batch.start_date && vote.date <= date_before_next_batch
```
instead of dates - use batch.starts_at

marcusbaguley commented on app/use_cases/create_voting_report.rb:
```diff
@@ -0,0 +1,50 @@
+  end
+
+  def perform
+    batches.sort_by! &:start_date
```
You could get the database to sort the batches when you retrieve them `Batch.where(..).order(starts_at: :asc)` 

It is a performance improvement - and in this case it probably is fine as the datasets are quite small, but in general good practice to do filtering and sorting in the database before the objects need to be loaded into ruby, as ruby is many times slower and hungry on memory than the database.

marcusbaguley commented on app/use_cases/determine_current_roast.rb:
```diff
@@ -0,0 +1,14 @@
+class DetermineCurrentRoast
+
```
I would not use `use cases` for these types of domain concepts. A service type object like a use case is useful when there is a long list of things to do that span multiple domain objects. In this case a static method on the Roast I think is more appropriate, easier to test, and find to use in other scenarios.

You could create a static method on the Roast class called `active_roast`

mattlee68 committed:
CC#4 Fix requested changes - vote now voted_at and ratings system has changed

mattlee68 committed:
CC#8 Fix requested changes
Batch.start_date is now Batch.starts_at, sorting is done more efficiently and removed use case for Current Batch

mattlee68 committed:
CC#8 Fix merge conflicts

mattlee68 committed:
CC#8 Fix retrieval of active roast

mattlee68 committed:
Merge branch 'feature/vote' into feature/count_votes

mattlee68 committed:
CC#8 Fix creation of vote in Vote Controller

mattlee68 committed:
CC#8 Remove admin helper
